### ISSUE TITLE
IPython notebook 3.0 pyzmq error on build for Linux Mint 17.1 - workaround described.

### ISSUE THREAD

**Issue by ruxi** [ipython/ipython#7999]
## TL;DR

For Linux Mint 17.1, **sudo pip install "ipython[notebook]"** spits out an error when building the dependency for pyzmq. Installing the zmq dependency seperately using **sudo apt-get install libzmq3-dev** fixes this. 

This was not described [in the docs](http://ipython.org/install.html). Might be a good idea to include it. Tested on a freshly installed OS using default python 2.7. 

---

## Long version

Linux Mint 17.1 was upgraded from a fresh install of Linux Mint 17 (from live boot usb).

First terminal command that were ran after upgrading was:

1\. sudo apt-get python-pip
2\. sudo pip install "ipython[notebook]"

The latter didn't work. 

```
ruxi@lifemint ~ $ pip install 'ipython[notebook]'

...
Downloading/unpacking pyzmq>=13 (from ipython[notebook])
  Downloading pyzmq-14.5.0.tar.gz (997kB): 997kB downloaded
  Running setup.py (path:/tmp/pip_build_ruxi/pyzmq/setup.py) egg_info for package pyzmq
    
    no previously-included directories found matching 'docs/build'
    no previously-included directories found matching 'docs/gh-pages'
    warning: no previously-included files found matching 'bundled/zeromq/src/Makefile*'
    warning: no previously-included files found matching 'bundled/zeromq/src/platform.hpp'
    warning: no previously-included files found matching 'setup.cfg'
    warning: no previously-included files found matching 'zmq/libzmq*'
    warning: no previously-included files matching '__pycache__/*' found anywhere in distribution
    warning: no previously-included files matching '.deps/*' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '.git*' found anywhere in distribution
    warning: no previously-included files matching '.DS_Store' found anywhere in distribution
    warning: no previously-included files matching '.mailmap' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.am' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.in' found anywhere in distribution
```
Pyzmq-14.5.0.tar.gz didn't build correctly on Linux Mint 17.1. 

Went to [pyzmq github page](https://github.com/zeromq/pyzmq) and it said:

> When compiling pyzmq (e.g. installing with pip on Linux), it is generally recommended that zeromq be installed separately, via homebrew, apt, yum, etc:

So I ran:

> sudo apt-get install libzmq3-dev

Which went something like this:

```
 ruxi@lifemint ~ $ sudo apt-get install libzmq3-dev

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following extra packages will be installed:
  libpgm-5.1-0 libzmq3
The following NEW packages will be installed:
  libpgm-5.1-0 libzmq3 libzmq3-dev
0 upgraded, 3 newly installed, 0 to remove and 4 not upgraded.
```

Ok. With pyzmq installed, I tried building ipython notebook again.
```
ruxi@lifemint ~ $ sudo pip install 'ipython[notebook]'

Downloading/unpacking pyzmq>=13 (from ipython[notebook])
  Downloading pyzmq-14.5.0.tar.gz (997kB): 997kB downloaded
  Running setup.py (path:/tmp/pip_build_root/pyzmq/setup.py) egg_info for package pyzmq
    
    no previously-included directories found matching 'docs/build'
    no previously-included directories found matching 'docs/gh-pages'
    warning: no previously-included files found matching 'bundled/zeromq/src/Makefile*'
    warning: no previously-included files found matching 'bundled/zeromq/src/platform.hpp'
    warning: no previously-included files found matching 'setup.cfg'
    warning: no previously-included files found matching 'zmq/libzmq*'
    warning: no previously-included files matching '__pycache__/*' found anywhere in distribution
    warning: no previously-included files matching '.deps/*' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '.git*' found anywhere in distribution
    warning: no previously-included files matching '.DS_Store' found anywhere in distribution
    warning: no previously-included files matching '.mailmap' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.am' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.in' found anywhere in distribution

...

  Running setup.py install for pyzmq
    Using zmq-prefix /usr (found via pkg-config).
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include -c build/temp.linux-x86_64-2.7/scratch/check_sys_un.c -o build/temp.linux-x86_64-2.7/scratch/check_sys_un.o
    x86_64-linux-gnu-gcc -pthread build/temp.linux-x86_64-2.7/scratch/check_sys_un.o -o build/temp.linux-x86_64-2.7/scratch/check_sys_un
    Configure: Autodetecting ZMQ settings...
        Custom ZMQ dir:
    ************************************************
    cc -c /tmp/timer_createmM4SrI.c -o build/temp.linux-x86_64-2.7/scratch/tmp/timer_createmM4SrI.o
    cc build/temp.linux-x86_64-2.7/scratch/tmp/timer_createmM4SrI.o -o build/temp.linux-x86_64-2.7/scratch/a.out
    build/temp.linux-x86_64-2.7/scratch/tmp/timer_createmM4SrI.o: In function `main':
    timer_createmM4SrI.c:(.text+0x15): undefined reference to `timer_create'
    collect2: error: ld returned 1 exit status
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -c build/temp.linux-x86_64-2.7/scratch/vers.c -o build/temp.linux-x86_64-2.7/scratch/vers.o
    x86_64-linux-gnu-gcc -pthread build/temp.linux-x86_64-2.7/scratch/vers.o -lzmq -lrt -o build/temp.linux-x86_64-2.7/scratch/vers
        ZMQ version detected: 4.0.4
    Warning: Detected ZMQ version: 4.0.4, but pyzmq targets ZMQ 4.0.5.
    Warning: libzmq features and fixes introduced after 4.0.4 will be unavailable.
    ************************************************
    ************************************************
    building 'zmq.devices.monitoredqueue' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/devices/monitoredqueue.c -o build/temp.linux-x86_64-2.7/zmq/devices/monitoredqueue.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/devices/monitoredqueue.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/devices/monitoredqueue.so
    building 'zmq.backend.cython._device' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/_device.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/_device.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/_device.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/_device.so
    building 'zmq.backend.cython._poll' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/_poll.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/_poll.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/_poll.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/_poll.so
    building 'zmq.backend.cython._version' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/_version.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/_version.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/_version.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/_version.so
    building 'zmq.backend.cython.constants' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/constants.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/constants.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/constants.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/constants.so
    building 'zmq.backend.cython.context' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/context.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/context.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/context.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/context.so
    building 'zmq.backend.cython.error' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/error.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/error.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/error.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/error.so
    building 'zmq.backend.cython.message' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/message.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/message.o
    zmq/backend/cython/message.c: In function ‘__pyx_pf_3zmq_7backend_6cython_7message_5Frame_24get’:
    zmq/backend/cython/message.c:3198:9: warning: variable ‘__pyx_v_property_c’ set but not used [-Wunused-but-set-variable]
       char *__pyx_v_property_c;
             ^
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/message.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/message.so
    building 'zmq.backend.cython.socket' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/socket.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/socket.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/socket.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/socket.so
    building 'zmq.backend.cython.utils' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_SYS_UN_H=1 -I/usr/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -I/usr/include/python2.7 -c zmq/backend/cython/utils.c -o build/temp.linux-x86_64-2.7/zmq/backend/cython/utils.o
    x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/zmq/backend/cython/utils.o -lzmq -lrt -o build/lib.linux-x86_64-2.7/zmq/backend/cython/utils.so
    
    no previously-included directories found matching 'docs/build'
    no previously-included directories found matching 'docs/gh-pages'
    warning: no previously-included files found matching 'bundled/zeromq/src/Makefile*'
    warning: no previously-included files found matching 'bundled/zeromq/src/platform.hpp'
    warning: no previously-included files found matching 'setup.cfg'
    warning: no previously-included files found matching 'zmq/libzmq*'
    warning: no previously-included files matching '__pycache__/*' found anywhere in distribution
    warning: no previously-included files matching '.deps/*' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '.git*' found anywhere in distribution
    warning: no previously-included files matching '.DS_Store' found anywhere in distribution
    warning: no previously-included files matching '.mailmap' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.am' found anywhere in distribution
    warning: no previously-included files matching 'Makefile.in' found anywhere in distribution
  Running setup.py install for ptyprocess
    
  Running setup.py install for certifi
    
  Running setup.py install for backports.ssl-match-hostname
    
Successfully installed ipython pyzmq jsonschema ptyprocess certifi backports.ssl-match-hostname
Cleaning up...
```

Its ugly, but it worked. :)

> ruxi@lifemint ~ $ ipython notebook
```
[I 00:56:12.290 NotebookApp] Using existing profile dir: u'/home/ruxi/.ipython/profile_default'
[I 00:56:12.294 NotebookApp] Writing notebook server cookie secret to /home/ruxi/.ipython/profile_default/security/notebook_cookie_secret
[I 00:56:12.295 NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
[I 00:56:12.315 NotebookApp] Serving notebooks from local directory: /home/ruxi
[I 00:56:12.316 NotebookApp] 0 active kernels 
[I 00:56:12.316 NotebookApp] The IPython Notebook is running at: http://localhost:8888/
[I 00:56:12.316 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
Created new window in existing browser session.
```

**Comment by minrk** [comment]
We don't enumerate the dependencies on the brief install page, but if you follow the "detailed instructions" link on that page, it takes you to a more detailed page where pyzmq is discussed.

Any pyzmq build failure would be a pyzmq issue, not an IPython one.

Can you provide the output of the failed build?
